pkg/dockerregistry/server/middlewares.go

```diff
@@ -0,0 +1,88 @@
+package server
+
+import (
+	"fmt"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/docker/distribution"
+	"github.com/docker/distribution/configuration"
+	"github.com/docker/distribution/context"
+	registryauth "github.com/docker/distribution/registry/auth"
+	"github.com/docker/distribution/registry/middleware/registry"
+	repomw "github.com/docker/distribution/registry/middleware/repository"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	registrystorage "github.com/docker/distribution/registry/storage/driver/middleware"
+)
+
+const (
+	AuthOpenShift = "openshift"
+
+	storageOpenShift = "openshift"
+	storageAppParam  = "__app__"
+)
+
+func patchConfig(config *configuration.Configuration, app *App) {
+	if config.Auth.Type() == AuthOpenShift {
+		if config.Auth[AuthOpenShift] == nil {
+			config.Auth[AuthOpenShift] = make(configuration.Parameters)
+		}
+		config.Auth[AuthOpenShift][AccessControllerOptionParams] = AccessControllerParams{
+			Logger:           context.GetLogger(context.Background()), // FIXME
+			SafeClientConfig: app.registryClient.SafeClientConfig(),
+			MetricsConfig:    app.extraConfig.Metrics,
+		}
+	}
+
+	for i := range config.Middleware["storage"] {
+		middleware := &config.Middleware["storage"][i]
+		if middleware.Name == storageOpenShift {
+			if middleware.Options == nil {
+				middleware.Options = make(configuration.Parameters)
+			}
+			middleware.Options[storageAppParam] = app
+		}
+	}
+}
+
+func init() {
+	registryauth.Register(AuthOpenShift, registryauth.InitFunc(newAccessController))
+
+	registrystorage.Register(storageOpenShift, func(driver storagedriver.StorageDriver, options map[string]interface{}) (storagedriver.StorageDriver, error) {
+		app, _ := options[storageAppParam].(*App)
+		if app == nil {
+			return nil, fmt.Errorf("failed to find an application instance in the options of the storage driver middleware")
+		}
+
+		log.Info("OpenShift middleware for storage driver initializing")
+
+		// We can do this because of an initialization sequence of middlewares.
+		// Storage driver is required to create registry. So we can be sure that
+		// this assignment will happen before registry and repository initialization.
+		app.driver = driver
+
+		return driver, nil
+	})
+
+	middleware.Register("openshift", func(ctx context.Context, registry distribution.Namespace, options map[string]interface{}) (distribution.Namespace, error) {
+		app := appFrom(ctx)
+		if app == nil {
+			return nil, fmt.Errorf("failed to find an application instance in the context of the registry middleware")
+		}
+
+		log.Info("OpenShift registry middleware initializing")
```

Comment:
Can we ditch the product name? E.g. replace the `OpenShift` with `downstream` or `origin` or just lowercased `openshift` to refer to a config section name instead of product.